Refactor TimingChange to make it easier to follow

smoogipoo opened the pull request:
Round 2.

Basic workflow is:
- A playfield will contain a `SpeedAdjustmentCollection`. `SpeedAdjustmentCollection` contains multiple (abstract) `SpeedAdjustmentContainer`s.
- A `SpeedAdjustmentContainer` creates and contains _one_ (abstract) `DrawableTimingChange`. This is the element that scrolls.
- A `DrawableTimingChange` contains many hit objects.

So to add hit objects you need to:
- Construct a `SpeedAdjustmentContainer` with an overriden `CreateTimingSection()`.
- Add the `SpeedAdjustmentContainer` to the `SpeedAdjustmentCollection`
- Add a `DrawableHitObject` to a `SpeedAdjustmentCollection`

TestCaseScrollingHitObjects is provided as an implementation demo.

Sound good? I made this generic enough so that it is usable in taiko (supports both y-relative and x-relative movement).

Does not currently support:

Adding new timing changes (after all the hit objects have been added).
Removing timing changes.
Reordering of timing changes.
Reordering of hit objects (between multiple timing changes).

smoogipoo committed:
Update to match framework.

smoogipoo committed:
Initial rewrite of timing changes to allow them to be more extensible.

smoogipoo committed:
Add bar line timing changes.

smoogipoo committed:
Move drawable timing changes to Timing/Drawables.

smoogipoo committed:
aaaaa

smoogipoo committed:
Fix incorrect sorting.

smoogipoo committed:
Add gravity timing change.

smoogipoo committed:
Cleanup + make travel time equal to the time span.

smoogipoo committed:
Set lifetime for timing change containers very naively for now.

smoogipoo committed:
Add gravity mod.

smoogipoo committed:
Add gravity mod icon.

smoogipoo committed:
Make default mania playfield timespan 1.5s for now.

smoogipoo committed:
Fix up test case.

smoogipoo committed:
Significantly reduce test case time.

smoogipoo committed:
Make life time set after children are updated (for now).

smoogipoo committed:
CI fixes.

smoogipoo committed:
Add missing license headers.

smoogipoo committed:
Super huge abstraction changes to (hopefully) make this TimingChangeContainer usable for non-mania rulesets.

Also includes commenting improvements/fixes.

smoogipoo committed:
Simplify gravity.

smoogipoo committed:
Add comment.

smoogipoo committed:
CI fixes.

smoogipoo committed:
Move timing change stuff to osu.Game project.

smoogipoo committed:
Merge branch 'master' into timingchange-rework

smoogipoo committed:
Properly set starting lifetime for mania hit objects.

smoogipoo committed:
Drastically improve mania gameplay loading time.

smoogipoo committed:
Update framework.

smoogipoo committed:
Merge remote-tracking branch 'origin/master' into timingchange-rework

# Conflicts:
#	osu.Game/Graphics/Containers/ReverseDepthFillFlowContainer.cs

smoogipoo committed:
Update with framework changes.

smoogipoo committed:
Update with framework changes.

smoogipoo committed:
Update with more framework changes.

smoogipoo committed:
Make hold notes utilize RelativeChildOffset for tick positionings.

peppy committed:
Merge remote-tracking branch 'upstream/master' into timingchange-rework

smoogipoo committed:
Make TimeSpan a double again.

smoogipoo committed:
Rename timing changes, simplify + improve columns/dictionary usage.

smoogipoo committed:
CI fixes.

smoogipoo committed:
g -> gravity + one more xmldoc.

smoogipoo committed:
Not sure how these CI issues were missed.

smoogipoo committed:
Remove IHasTimeSpan, add rename to TimingSection.

smoogipoo committed:
Split out RelativeCoordinateAutoSizingContainer into HitObjectCollection, beginning major rework/refactoring.

smoogipoo committed:
More renaming + fixes.

smoogipoo committed:
Rewrite mania to use the new timing section stuff.

smoogipoo committed:
More cleanup + xmldoc changes.

smoogipoo committed:
Merge remote-tracking branch 'origin/master' into timingchange-rework

# Conflicts:
#	osu-framework
#	osu.Game.Rulesets.Mania/Timing/ControlPointContainer.cs

smoogipoo committed:
Update framework.

peppy committed:
More xmldoc

smoogipoo committed:
Use new naming structure + VisibleTimeRange bindable.

smoogipoo committed:
Namespace movement.

smoogipoo committed:
Rename SpeedAdjustment -> MultiplierControlPoint + reworking.

smoogipoo committed:
Rewrite the way speed adjustments are applied.

smoogipoo committed:
Make bar lines work again.

smoogipoo committed:
Merge remote-tracking branch 'origin/master' into timingchange-rework

# Conflicts:
#	osu-framework

smoogipoo committed:
Cleanups + xmldoc rewordings/improvements.

smoogipoo committed:
Add small test case to demonstrate usage.

smoogipoo committed:
Remove unused using.

peppy commented:
On a first pass, definitely looking a lot better! I think the values for the testcase could be tweaked a bit (and maybe add labels to show the visibletimerange etc.)

peppy committed:
Tidy up testcase a bit

peppy committed:
Set a more sane default for TimingControlPoint's BeatLength

peppy committed:
Add xmldoc to MultiplierControlPoint

peppy committed:
Use a relative width definition

peppy committed:
Add labels to playfield area

peppy committed:
Add a check to make sure all hitobjects use relativepositionaxes

peppy committed:
Enforce exception better

peppy committed:
Add current time labels

peppy committed:
Make hitobjects a bit more visual

smoogipoo committed:
Add comment.

smoogipoo committed:
Merge branch 'master' into timingchange-rework

smoogipoo committed:
Move speed adjustment code from ManiaHitRenderer to SpeedAdjustedHitRenderer.

smoogipoo committed:
Make SpeedAdjustmentCollection support unordered adds of speed adjustments and hit objects.

smoogipoo committed:
We don't need this to be protected anymore.

smoogipoo committed:
A bit more cleanup.

smoogipooo commented:
@peppy made a few changes to reduce implementation complexity, let me know if you like it.

peppy commented on osu.Game/Rulesets/Timing/DrawableTimingSection.cs:
```diff
@@ -0,0 +1,111 @@
+
+        private readonly Axes autoSizingAxes;
+
+        private Cached layout = new Cached();
```
any reason for this? can't just check if queue is empty?

smoogipoo committed:
We don't need to use Cached just yet.

This was here to support dynamic re-ordering of hit objects and speed adjustment containers, however there needs to be more work done to TimingControlPoints/MultiplierControlPoints in general before this is possible.

smoogipoo committed:
CI fixes.

Update SpeedAdjustmentCollection.cs

smoogipoo committed:
Remove gross autoSizingAxes and life time.

smoogipoo committed:
Implement proper expiry in test case.

smoogipoo committed:
CI fixes.

smoogipoo committed:
We no longer need to pass in the control point.

peppy committed:
Merge branch 'master' into timingchange-rework
